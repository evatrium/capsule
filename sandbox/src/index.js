let log_this = '';
const log = (it) => {
    let val;
    if (typeof it !== 'string') {
        val = JSON.stringify(it, null, 4);
    } else {
        val = it;
    }
    log_this = '\n' + log_this + '\n' + val + '\n';
};


export const tester = () => {
    return new Promise((resolve, reject) => {
        console.log(log_this);

        console.log('working');


        resolve()
    })
};