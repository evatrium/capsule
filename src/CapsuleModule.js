import React, {Component} from "react";

class Root extends Component {
    render() {
        return (
            <React.Fragment>
                {this.props.children}
            </React.Fragment>
        )
    }
}

export const CapsuleModule = ({createStore, actionsCreator, classNames, asyncHOC, connect, Provider, jssProcessor, events}) => {

    const store = createStore();
    const cn = classNames;
    let collective = {};
    let actions = {};
    let theme = {};
    let propsFromProvider = {};
    let logicFromProvider = {};
    let _jssProcessor = jssProcessor;
    let stylesHOC;
    let RootComponent;

    const CapsuleProvider = (config = {}) => {

        theme = config.theme || {};
        propsFromProvider = config.toProps || {};
        logicFromProvider = config.toLogic || {};
        _jssProcessor = config.stylesHOC ? false : config.jssProcessor || jssProcessor;
        stylesHOC = config.stylesHOC;
        RootComponent = config.RootComponent || Root;
        collective = {...collective, ...config.toCollective};


        return (Child = null) => {

            const AC = asyncHOC && config.loadApp ? asyncHOC(config.loadApp) : null;

            const App = AC || Child;

            return () => (
                <Provider
                    store={store}
                    collective={{getCollective: () => collective}}
                    actions={{getActions: () => actions}}>
                    <RootComponent>
                        <App/>
                    </RootComponent>
                </Provider>
            );
        }
    };

    const makeActionsAndAttachState = (initialState, actionsName) => {
        if (!initialState || !actionsName) return;
        store.attachState({[actionsName]: initialState});
        actions = actionsCreator(actionsName, initialState, store, actions);
    };

    const makeLogic = (logic) => {
        if (!logic) return {};
        return logic({
            ...logicFromProvider,
            events,
            actions,
            store,
            collective: () => collective
        })
    };

    const updateCollective = (logicName, returnLogic) => {
        if (!logicName) return;
        collective = {
            ...collective,
            [logicName]: returnLogic
        }
    };

    const makeClassesFromStyles = (styles) => {
        if (!styles || !_jssProcessor) return null;
        return _jssProcessor(typeof styles === 'function' ? styles(theme) : styles);
    };

    const Capsule = (c = {}) => {

        let {
            name,
            styles,
            initialState,
            actionsName,
            mapState,
            mapActions,
            logicName,
            logic,
            mapLogic,
        } = c;

        if (name) {
            if (!actionsName) actionsName = name;
            if (!logicName) logicName = name;
        }

        makeActionsAndAttachState(initialState, actionsName);

        let returnLogic = makeLogic(logic);

        updateCollective(logicName, returnLogic);

        let classes = makeClassesFromStyles(styles);

        return (Child) => {
            if (!Child) return returnLogic;

            if (styles && stylesHOC) {
                Child = stylesHOC(styles)(Child);
            }

            return connect({
                mapState,
                mapActions,
                mapLogic,
                additionalProps: {
                    cn,
                    theme,
                    ...propsFromProvider,
                    classes
                }
            })(Child)
        };
    };

    return {
        CapsuleProvider,
        Capsule,
    }

};
