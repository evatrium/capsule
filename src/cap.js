import {Capsule} from "./index";


const user = Capsule('user', {
    active: false,
    first_name: '',
    last_name: '',
    src: ''
});

const onUserStateChange = ({active}) => {
    user.merge({active});
};

const Avatar = user(({active, first_name, last_name, src}) => {

    return (
        <UserBox>
            <Img src={src}/>
            <Typo text={first_name + ' ' + last_name}/>
        </UserBox>
    );
});

