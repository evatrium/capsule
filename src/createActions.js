import {basicNestedCopy} from "./utils";

export const createActions = (actionsName, initialState, store, actions, noAdditionalImmutability) => {
    initialState = basicNestedCopy(initialState, noAdditionalImmutability);
    //set up the placeholders for all the actions on the namespace
    actions[actionsName] = {set: {}, get: {}, toggle: {}, update: () => null, merge: () => null};

    actions[actionsName].update = (cb, skipCopy) => {
        let state = store.getState();
        // pass the namespace state to the callback,
        // and assign the returned value to the namespace
        store.setState({
            ...state,
            [actionsName]: basicNestedCopy(cb({...state[actionsName]}), noAdditionalImmutability || skipCopy)
        });
    };

    actions[actionsName].merge = (propsToMerge, skipCopy) => {
        let state = store.getState();
        //spread the propsToMerge and assign to the namespace
        store.setState({
            ...state,
            [actionsName]: {
                ...state[actionsName],
                ...basicNestedCopy(propsToMerge, noAdditionalImmutability || skipCopy)
            }
        });
    };

    //return the namespace state
    actions[actionsName].getState = (skipCopy) =>
        basicNestedCopy(store.getState()[actionsName], noAdditionalImmutability || skipCopy);

    //for every key in the state
    //set the value as the function name of the action type
    //{foo:1} ... namespace.set.foo(2);
    Object.keys(initialState).forEach(key => {

        //set a single value onto the namespace
        actions[actionsName].set[key] = (newValue, skipCopy) => {
            let state = store.getState();
            store.setState({
                ...state,
                [actionsName]: {
                    ...state[actionsName],
                    [key]: basicNestedCopy(newValue, noAdditionalImmutability || skipCopy)
                }
            });
        };

        //get a single value from the namespace
        actions[actionsName].get[key] = (skipCopy) => {
            return basicNestedCopy(store.getState()[actionsName][key], noAdditionalImmutability || skipCopy);
        };

        //toggle a single value on a namespace to its inverse
        actions[actionsName].toggle[key] = () => {
            actions[actionsName].set[key](!actions[actionsName].get[key]())
        };
    });

    return actions;
};
