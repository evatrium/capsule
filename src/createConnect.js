import React, {Component, createElement} from 'react';
import {convertArguments} from "./utils";

/**
 * selects the values off of the state
 * accepts a function to map props like this:
 * (state) => ({mySingleProp: state.namespace.prop});
 * or with comma separated values on an object like this:
 * {namespace:'prop,someOtherProp', someOtherNameSpace: 'foo,bar'}
 * @param {function|Object} mapper - plucks the keys off of the state
 * @param {Object} source - the state
 * @returns {Object} - the selected state
 */
export const mapIt = (mapper, source) => {
    if (typeof mapper === 'function') return mapper(source);
    if (typeof mapper === 'object') {
        let selected = {};
        let keys = Object.keys(mapper);
        for (let i = 0; i < keys.length; i++) {
            const namespace = keys[i];
            if (!source[namespace]) break;
            let props = mapper[namespace];
            if (typeof props === 'string') props = props.split(/\s*,\s*/);
            for (let i = 0; i < props.length; i++) selected[props[i]] = source[namespace][props[i]];
        }
        return selected;
    }
    return {};
};

/*** Wire a component up to the store. Passes state as props, re-renders on change.
 * inspiration: https://github.com/developit/unistore/blob/master/src/integrations/react.js
 * @param CapsuleContext
 * @returns {function}*/
export const createConnect = (CapsuleContext) => (...args) => Child => {

    const {mapState, mapActions, mapLogic} = convertArguments(...args);

    function Connected(props, context) {
        Component.call(this, props, context);
        context = context || {};
        const {store, getCollective, getActions} = context;

        let mappedProps = {}, state = {},
            getState = () => store && store.getState ? store.getState() : {},
            update = () => {
            }, sub, unsub;

        //map the collections to props
        if (mapLogic && getCollective) mappedProps = mapIt(mapLogic, getCollective());
        if (mapActions && getActions) mappedProps = {...mappedProps, ...mapIt(mapActions, getActions())};
        if (mapState) { //skip if there is no mapState in case we are just mapping logic or actions
            state = mapIt(mapState, getState()); //first run-through, map the state
            update = () => {
                let mapped = mapIt(mapState, getState()); //get the plucked state values onto an object
                for (let i in mapped) if (mapped[i] !== state[i]) {//and compare each value with the last
                    state = mapped; // if there is a change then update the state
                    //we can dip out at this point. check if mounted so we dont update after unmounting
                    return this.mounted && this.forceUpdate();//simulate setState on the wrapped component
                }// if there aren't any updates then check to see if any new properties have been added
                for (let i in state) if (!(i in mapped)) {
                    state = mapped;
                    return this.mounted && this.forceUpdate();
                }
            }; //subscribe the update function to be called whenever store.setState is called
            sub = () => store && store.subscribe(update);
            unsub = () => store && store.unsubscribe(update);
        }

        this.componentDidMount = () => {
            this.mounted = true;
            update(); //need to call update here
            //in case the component (that this wraps) triggers an update
            //prior to subscribing to the store.
            sub && sub();
        };
        this.componentWillUnmount = () => {
            this.mounted = false;
            unsub && unsub();
        };
        this.render = () =>
            createElement(Child, {...mappedProps, ...state, ...this.props});
    }

    const ConnectedCapsule = (Connected.prototype = new Component()).constructor = Connected;
    ConnectedCapsule.contextType = CapsuleContext;
    return ConnectedCapsule;
};
