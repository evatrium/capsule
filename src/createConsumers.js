import {createElement, useContext, useReducer, useRef, useMemo, useEffect} from 'react';

export const mapIt = (mapper, source) => {
    if (typeof mapper === 'function') return mapper(source);
    if (typeof mapper === 'object') {
        let selected = {};
        let keys = Object.keys(mapper);
        for (let i = 0; i < keys.length; i++) {
            const namespace = keys[i];
            if (!source[namespace]) break;
            let props = mapper[namespace];
            if (typeof props === 'string') props = props.split(/\s*,\s*/);
            for (let i = 0; i < props.length; i++) selected[props[i]] = source[namespace][props[i]];
        }
        return selected;
    }
    return {};
};

export const createConsumers = (CapsuleContext) => {

    const useForceUpdate = () => useReducer(state => !state, false)[1];

    const useCapsule = ({mapState, mapActions, mapLogic}) => {
        const {store, getCollective, getActions} = useContext(CapsuleContext);
        let forceUpdate = useForceUpdate(),
            ignoreForce = useRef(false),
            hasMounted = useRef(false),
            update = useRef(() => null),
            state = useRef(store.getState()),
            mappedProps = useRef({});

        mappedProps.current = useMemo(() => {
            let m_props = {};
            if (mapLogic && getCollective) m_props = mapIt(mapLogic, getCollective());
            if (mapActions && getActions) m_props = {...m_props, ...mapIt(mapActions, getActions())};
            return m_props;

        }, []);

        const updater = useMemo(() => (_, initial) => {
            let mapped = mapIt(mapState, store.getState());
            for (let i in mapped) if (mapped[i] !== state.current[i]) {
                state.current = mapped;
                return !ignoreForce.current && !initial && forceUpdate()
            }
            for (let i in state.current) if (!(i in mapped)) {
                state.current = mapped;
                return !ignoreForce.current && !initial && forceUpdate()
            }

        }, []);

        if (!hasMounted.current && mapState) updater(null, true);

        useEffect(() => {
            hasMounted.current = true;
            if (mapState) {
                update.current = updater;
                update.current();
            }
            const unsub = store.subscribe(update.current);
            return () => {
                ignoreForce.current = true;
                unsub();
            }

        }, []);

        return [state.current, mappedProps.current]
    };

    const connectCapsule = ({mapState, mapActions, mapLogic}) => Child => props => {
        const [state, mappedProps] = useCapsule({mapState, mapActions, mapLogic});
        return createElement(Child, {...mappedProps, ...state, ...props});
    };

    return {
        useCapsule,
        connectCapsule,
    }
};
