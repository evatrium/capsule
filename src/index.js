export {
    CapsuleProvider,
    Capsule,
    events,
    lazyLoader,
    routing,
    Router,
    Linkage
} from './created';