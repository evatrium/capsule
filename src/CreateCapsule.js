import {createStore, createActions, connect, Provider} from "@iosio/store";
import {classNames} from "@iosio/utils/lib/classNames";
import {lazyLoader } from "./lazyLoader";
import {CapsuleModule} from "./CapsuleModule";
import {JSS_nano} from "@iosio/jss_nano";
import {Eventer} from '@iosio/utils/lib/eventer';
import {createRouting} from "./routing";

export const CreateCapsule = () => {

    const events = Eventer();

    const jss_nano = JSS_nano(process.env.NODE_ENV === 'development');

    const {CapsuleProvider, Capsule} = CapsuleModule({
        createStore,
        connect,
        Provider,
        actionsCreator: createActions,
        classNames,
        asyncHOC: lazyLoader,
        jssProcessor: jss_nano,
        events
    });

    const {
        routing,
        Router,
        WithRouting,
        Linkage
    } = createRouting(Capsule);

    return {
        CapsuleProvider,
        Capsule,
        events,
        lazyLoader,
        routing,
        Router,
        Linkage,
        WithRouting,
        jss_nano
    }

};
