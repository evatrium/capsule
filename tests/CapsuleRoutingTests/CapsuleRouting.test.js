import React from 'react'
import {CapsuleProvider, Capsule, Router, Linkage, lazyLoader, routing, events} from '../../src';
import renderer from 'react-test-renderer';
import {render, fireEvent, queryByAttribute} from 'react-testing-library';
import {Till} from "../_test_utils";
import {mainCapsule} from "./mainCapsule";
import Home from '../_test_utils/pages/Home';
import Detail from '../_test_utils/pages/Detail';
import AuthOnly from '../_test_utils/pages/AuthOnly';
import Restricted from '../_test_utils/pages/Restricted';
import Login from '../_test_utils/pages/Login';

const authControls = mainCapsule();


const till = Till(events);

const getById = queryByAttribute.bind(null, 'id');

const {goTo, getLocation} = routing;


// const Detail = lazyLoader(() => import('../_test_utils/pages/Detail'));
// const AuthOnly = lazyLoader(() => import('../_test_utils/pages/AuthOnly'));
// const Restricted = lazyLoader(() => import('../_test_utils/pages/Restricted'));
// const Login = lazyLoader(() => import('../_test_utils/pages/Login'));

const Link = ({toPath, toParams, name, id}) => (
    <Linkage toPath={toPath} toParams={toParams} id={id} className={'link'}>
        {({pathname}) => {
            const activeClass = toPath === pathname ? 'currPath' : null;
            return (
                <div className={activeClass}>
                    {name}
                </div>
            );
        }}
    </Linkage>
);



const Nav = Capsule({
    mapLogic: {main: 'setAdmin,logout,login'},
    mapState: {main: 'loggedIn,isAdmin'},
})(({onSubmitText, text, loggedIn, logout, setAdmin, login, isAdmin}) => (
    <nav className={'nav'}>

        <div className={'flexRow'}>

            <Link toPath={'/'} id={'home'} name={'home'}/>
            <Link toPath={'/detail'} id={'detail'} name={'detail'}/>
            <Link toPath={'/authOnly'} id={'authOnly'} name={'authOnly'}/>
            <Link toPath={'/restricted'} onClick={() => {
                cosole.log('restricted clicked')
            }} id={'restricted'} name={'restricted'}/>

        </div>

        <div className={'flexRow'}>
            {loggedIn ?
                <button className={'btn'} id="logout" onClick={logout}>Logout</button>
                : <button className={'btn'} id="login" onClick={login}>login</button>
            }
            <button id="toggleAdmin" className={'btn'} onClick={() => setAdmin()}>
                is Admin: {isAdmin ? "true" : "false"}
            </button>
        </div>

    </nav>
));


const WithRoutingCapsule = Capsule({
    mapLogic: {main: 'login'},
    mapState: {main: 'loggedIn,isAdmin'},
})(class WithRoutingCapsule extends React.Component {

    render() {

        const {login, loggedIn, isAdmin} = this.props;

        const publicRouts = {
            '/': Home,
            '/detail': Detail,
            '/login': Login,

        };

        const protectedRoutes = {
            '/authOnly': AuthOnly,
            '/restricted': Restricted,
        };

        let pathMap = {
            ...publicRouts
        };

        if (loggedIn) {
            pathMap = {
                ...publicRouts,
                ...protectedRoutes
            }
        }

        const accessiblePaths = Object.keys(pathMap);

        return (
            <React.Fragment>

                <Nav/>

                <Router
                    accessiblePaths={accessiblePaths}
                    noMatch={'/'}
                    pathMap={pathMap}/>

            </React.Fragment>
        );
    }
});

const createApp = () => CapsuleProvider()(WithRoutingCapsule);

describe("routing with react test renderer", () => {

    it('renders without crashing', () => {


        const App = createApp();

        const tree = renderer.create(<App/>);

        expect(tree).toMatchSnapshot();


    });

});


describe('it renders the correct component when the location is changed', () => {


    const App = createApp();

    const {container} = render(<App/>);

    const clickIt = (id) => {
        fireEvent.click(getById(container, id));
    };

    const runTest = ({click, startUrl, expected = {}, cb}) =>
        new Promise(resolve => {
            const {endUrl, endName} = expected;


            if (click === 'restricted') {
                // console.log(getLocation().url)
            }


            startUrl && expect(getLocation().url).toBe(startUrl);


            let resolved = false;


            let timeout

            const callback = ({name}) => {
                if (!resolved) {
                    setTimeout()
                    console.log(name)
                    expect(name).toBe(endName);
                    expect(getLocation().url).toBe(endUrl);
                    resolve();
                    cb && cb()
                    resolved = true;
                }
                events.off('mounted', callback);
                clearTimeout(timeout);
            };

            events.on('mounted', callback);

            click && clickIt(click);

            if (!resolved) {

                timeout = setTimeout(() => {
                    if (!resolved) {
                        resolve();
                        expect(getLocation().url).toBe(endUrl);
                    }
                    events.off('mounted', callback);
                }, 1000);

            }

        });


    it('renders the correct route when routed to', () => {

        return runTest({
            startUrl: '/',
            click: 'detail',
            expected: {
                endUrl: '/detail',
                endName: 'Detail'
            }
        });

    });


    it('it does not go to a protected route when routed to', () => {

        // expect.assertions(3);
        return runTest({
            startUrl: '/detail',
            click: 'authOnly',
            expected: {
                endUrl: '/detail',
            }
        });

    });


    // it('it allows a protected route to be accessible upon login', () => {
    //
    //     // expect.assertions(3);
    //     return runTest({
    //         startUrl: '/',
    //         click: 'login',
    //         expected: {
    //             endName: 'AuthOnly',
    //             endUrl: '/authOnly',
    //         }
    //     })
    // });


    // it('it allows another protected route to be accessible when logged in', () => {
    //
    //     // expect(authControls.getLoggedIn()).toBe(true);
    //
    //
    //     return new Promise((resolve=>{
    //
    //         return runTest({
    //             click: 'restricted',
    //             startUrl: '/authOnly',
    //             cb: resolve,
    //             expected: {
    //                 endName: 'Restricted',
    //                 endUrl: '/restricted',
    //             }
    //         });
    //     }))
    // });


    // it('it can navigate back home', () => {
    //     return runTest({
    //         startUrl: '/authOnly',
    //         click: 'home',
    //         expected: {
    //             endName: 'Home',
    //             endUrl: '/',
    //         }
    //     });
    // });
    //
    //
    // it('it can navigate back to the login page when logout is clicked', () => {
    //     return runTest({
    //         startUrl: '/',
    //         click: 'logout',
    //         expected: {
    //             endName: 'Login',
    //             endUrl: '/login',
    //         }
    //     }).then(() => {
    //             expect(authControls.getState().loggedIn).toBe(false);
    //             return runTest({
    //                 startUrl: '/login',
    //                 click: 'authOnly',
    //                 expected: {
    //                     endUrl: '/login',
    //                 }
    //             })
    //         }
    //     );
    //
    // });

});