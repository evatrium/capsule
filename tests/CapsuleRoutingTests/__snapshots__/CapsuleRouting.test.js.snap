// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routing with react test renderer renders without crashing 1`] = `
Array [
  <nav
    className="nav"
  >
    <div
      className="flexRow"
    >
      <a
        className="link"
        classes={null}
        href="/"
        id="home"
        onClick={[Function]}
        theme={Object {}}
      >
        <div
          className="currPath"
        >
          home
        </div>
      </a>
      <a
        className="link"
        classes={null}
        href="/detail"
        id="detail"
        onClick={[Function]}
        theme={Object {}}
      >
        <div
          className={null}
        >
          detail
        </div>
      </a>
      <a
        className="link"
        classes={null}
        href="/authOnly"
        id="authOnly"
        onClick={[Function]}
        theme={Object {}}
      >
        <div
          className={null}
        >
          authOnly
        </div>
      </a>
      <a
        className="link"
        classes={null}
        href="/restricted"
        id="restricted"
        onClick={[Function]}
        theme={Object {}}
      >
        <div
          className={null}
        >
          restricted
        </div>
      </a>
    </div>
    <div
      className="flexRow"
    >
      <button
        className="btn"
        id="login"
        onClick={[Function]}
      >
        login
      </button>
      <button
        className="btn"
        id="toggleAdmin"
        onClick={[Function]}
      >
        is Admin: 
        false
      </button>
    </div>
  </nav>,
  <div
    className="page"
  >
    {
    "name": "Home",
    "theme": {},
    "classes": null
}
  </div>,
  " ",
]
`;
